### PR TITLE
Deprecate sequtils.mapLiterals

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Add 1 `runnableExamples` to show `mapLiterals` only work on "Atomic" literals.

Got me by surprise because is not a `template` but a `macro` and **it wont map to all items**.

See https://github.com/nim-lang/Nim/pull/18577#issuecomment-886231715

### CODE REVIEW

**Comment by timotheecour** [comment]
what's the rationale for this API? it looks error-prone; can't we replace it with something more intuitive?

**Comment by juancarlospaco** [comment]
@timotheecour 
I did NOT created the API, I dunno. 🤷

@Araq done it, with the comment for ["easy construction of tuples"](https://github.com/nim-lang/Nim/commit/908a25a2ca262b212ab1b9e4dc4da42ece46b6df#diff-5bcacc6f2c5664f94022187523f52ec45aff40c875f75910a538512ac45428d8R722) 🤔

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -1096,8 +1096,10 @@ macro mapLiterals*(constructor, op: untyped;
   ## be used to map every array element to some target type:
   runnableExamples:
     let x = mapLiterals([0.1, 1.2, 2.3, 3.4], int)
-    doAssert x is array[4, int]
-    doAssert x == [int(0.1), int(1.2), int(2.3), int(3.4)]
+    assert x is array[4, int]
+    assert x == [int(0.1), int(1.2), int(2.3), int(3.4)]
+    ## op applies to atomic literals only
```
Comment (ringabout) [inline lib/pure/collections/sequtils.nim]: nitpick:
```suggestion
    # op applies to atomic literals only
```

**Comment by juancarlospaco** [comment]
`{.deprecated.}` ?. 🤔

**Comment by timotheecour** [comment]
IMO we should deprecate it for following reasons:
* it's not an intuitive API (changing only literals (but not const expressions) makes little sense)
* it says it's only changing literals but it's not true; it also changes some expressions that const-fold
```nim
when true:
  import sequtils
  const a = 1
  let x1 = mapLiterals((1, 2, a), int8)
  let x2 = mapLiterals((1, 2, 3*2), int8)
  echo x1.type # (int8, int8, int)
  echo x2.type # (int8, int8, int8)
```
* it's not flexible enough, eg:
```nim
when true:
  import sequtils
  let x = mapLiterals([(1, 2, 'a'), (3, 4, 'b')], int8)
  echo x
```
prints `[(1, 2, 97), (3, 4, 98)]` (there's no way to exclude char literals being transformed)

instead, we could use `mapIt` which doesn't have the above mentioned issues and is more flexible:
```nim
when true:
  import sequtils
  let x = mapIt(@[(1, 2, 'a'), (3, 4, 'b')], (it[0].int8, it[1].int16, it[2]))
  echo x
  echo x.type
```
prints:
```
@[(1, 2, 'a'), (3, 4, 'b')]
seq[(int8, int16, char)]
```

## note
there seems to be a bug in mapIt:
case6 works, case6b fails with: `Error: cannot evaluate at compile time: i`gensym0`
```nim
when defined case6:
  import sequtils
  const a = mapIt(@[1,2], it.uint8)
  echo a

when defined case6b:
  import sequtils
  proc main =
    const a = mapIt(@[1,2], it.uint8)
    echo a
  main()
```
(EDIT: that's just https://github.com/nim-lang/RFCs/issues/276 all over again)

File: lib/pure/collections/sequtils.nim [hunk 2]
```diff
@@ -1096,8 +1096,10 @@ macro mapLiterals*(constructor, op: untyped;
   ## be used to map every array element to some target type:
   runnableExamples:
     let x = mapLiterals([0.1, 1.2, 2.3, 3.4], int)
-    doAssert x is array[4, int]
-    doAssert x == [int(0.1), int(1.2), int(2.3), int(3.4)]
+    assert x is array[4, int]
+    assert x == [int(0.1), int(1.2), int(2.3), int(3.4)]
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim:1100]: IMO we should at least mention an example like this:
```nim
let x = mapLiterals([(1, 2), (3, 4)], int8)
assert x.type == ...
assert x == ...
```
which at least shows some usefulness for this feature (unlike the other examples given)

**Comment by juancarlospaco** [comment]
It is in fact a lottery whether it works or not, not intuitive at all.

**Comment by timotheecour** [comment]
I'm pretty sure it's deterministic, but the problem is that it's arbitrary/surprising and not flexible enough to be useful

File: lib/pure/collections/sequtils.nim [hunk 3]
```diff
@@ -1113,6 +1115,8 @@ macro mapLiterals*(constructor, op: untyped;
     assert d == ("1", (2, 3), "4", (5, 6))
   ## There are no constraints for the `constructor` AST, it
   ## works for nested tuples of arrays of sets etc.
+  ##
+  ## .. warning:: Deprecated; See https://github.com/nim-lang/Nim/pull/18577
```
Comment (konsumlamm) [inline lib/pure/collections/sequtils.nim]: This isn't necessary when it's already marked as `{.deprecated.}`, is it?

File: lib/pure/collections/sequtils.nim [hunk 4]
```diff
@@ -1090,14 +1090,16 @@ func mapLitsImpl(constructor: NimNode; op: NimNode; nested: bool;
         result.add v
 
 macro mapLiterals*(constructor, op: untyped;
-                   nested = true): untyped =
+                   nested = true): untyped {.deprecated.} =
```
Comment (timotheecour) [inline lib/pure/collections/sequtils.nim]:
```suggestion
                   nested = true): untyped {.deprecated: "use `mapIt` or other APIs instead".} =
```

File: lib/pure/collections/sequtils.nim (hunk 2 above)
Comment (timotheecour) [reply lib/pure/collections/sequtils.nim:1100]: what about this ?

Comment (juancarlospaco) [reply lib/pure/collections/sequtils.nim:1100]: Better to add more examples on non-Deprecated procs that are not footguns. 😆

**Comment by Araq** [comment]
So, if accepted, 1.6 deprecates `mapLiterals` because @juancarlospaco had trouble in using it. So eventually everybody who used `mapLiterals` successfully should be punished and rewrite some code. I don't think this is an acceptable strategy for developing Nim. We have an RFC process for these reasons, the process shields us from arbitrary "micro-improvements" that are more harmful than helpful.

**Comment by juancarlospaco** [comment]
Dont you want a less buggy Nim ?. 🤷

**Comment by Araq** [comment]
No, I don't, if your definition of "bug" is "stuff I'm confused about".

**Comment by juancarlospaco** [comment]
[This chainsaw juggling](https://github.com/nim-lang/Nim/pull/18577#issuecomment-886231715) is the feature that will make people migrate from Rust?, is it a better Pascal?.

**Comment by timotheecour** [comment]
see https://github.com/nim-lang/Nim/pull/18603 instead

**Comment by Araq** [comment]
> This chainsaw juggling is the feature that will make people migrate from Rust?, is it a better Pascal?.

Silly nonsense. Just try to get something deprecated over in Rust land, pick any `fn` over there that is confusing to you and see how far you get.

**Pull request closed without merging**